### PR TITLE
Hard code segment constants for common type sizes

### PR DESCRIPTION
**Pull request by sharwell**
* Fix calculation of array sizes
* Allow for inlining for common sizes on .NET Framework
* Reduces inline code size on .NET prior to version 8 (where lzcnt constant folding is first implemented)

Follow-up to #67558

### CODE REVIEW

**Review by EgorBo — APPROVED** [review]

File: src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs [hunk 1]
```diff
@@ -19,52 +19,64 @@ internal static class SegmentedArrayHelper
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         internal static int GetSegmentSize<T>()
         {
+            return Unsafe.SizeOf<T>() switch
+            {
+                4 => 16384,
```
Comment (EgorBo) [inline src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs:24]: Perhaps worth a comment explaining that these are just the most popular values?

Comment (sharwell) [reply src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs:26]: ➡️ Comments added

File: src/Compilers/Core/CodeAnalysisTest/Collections/SegmentedArrayHelperTests.cs [hunk 2]
```diff
@@ -12,6 +17,89 @@ namespace Microsoft.CodeAnalysis.UnitTests.Collections
...
+        [StructLayout(LayoutKind.Sequential, Size = 28)]
+        private struct Size28 { }
+
+        [StructLayout(LayoutKind.Sequential, Size = 32)]
+        private struct Size32 { }
+
+        [StructLayout(LayoutKind.Sequential, Size = 40)]
+        private struct Size40 { }
+
+        public static IEnumerable<object[]> ExplicitSizeTypes
```
Comment (AlekseyTs) [inline src/Compilers/Core/CodeAnalysisTest/Collections/SegmentedArrayHelperTests.cs:47]: >object [](http://example.com/codeflow?start=34&length=6)

Can we use more specific element type?

Comment (sharwell) [reply src/Compilers/Core/CodeAnalysisTest/Collections/SegmentedArrayHelperTests.cs:47]: This is the signature for properties referenced by `[MemberData]`:
https://github.com/xunit/xunit/blob/2f51d3627a5df6bcdc9ca9b17764cac06a7d6488/src/xunit.core/MemberDataAttribute.cs#L13

File: src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs [hunk 3]
```diff
@@ -19,52 +19,70 @@ internal static class SegmentedArrayHelper
...
+                16 => 4096,
+                24 => 2048,
+                28 => 2048,
+                32 => 2048,
+                40 => 2048,
 #if NETCOREAPP3_0_OR_NEWER
-            return InlineCalculateSegmentSize(Unsafe.SizeOf<T>());
+                _ => InlineCalculateSegmentSize(Unsafe.SizeOf<T>()),
 #else
-            if (Unsafe.SizeOf<T>() == Unsafe.SizeOf<object>())
```
Comment (AlekseyTs) [inline src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs:25]: >Unsafe.SizeOf<T>() == Unsafe.SizeOf<object>() [](http://example.com/codeflow?start=16&length=45)

Consider adding an assert verifying that this condition is always false.

File: src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs [hunk 4]
```diff
@@ -19,52 +19,70 @@ internal static class SegmentedArrayHelper
...
+                16 => 12,
+                24 => 11,
+                28 => 11,
+                32 => 11,
+                40 => 11,
 #if NETCOREAPP3_0_OR_NEWER
-            return InlineCalculateSegmentShift(Unsafe.SizeOf<T>());
+                _ => InlineCalculateSegmentShift(Unsafe.SizeOf<T>()),
 #else
-            if (Unsafe.SizeOf<T>() == Unsafe.SizeOf<object>())
```
Comment (AlekseyTs) [inline src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs:42]: >if (Unsafe.SizeOf<T>() == Unsafe.SizeOf<object>()) [](http://example.com/codeflow?start=12&length=50)

Same suggestion about an assert.

File: src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs [hunk 5]
```diff
@@ -19,52 +19,70 @@ internal static class SegmentedArrayHelper
...
+                16 => 4095,
+                24 => 2047,
+                28 => 2047,
+                32 => 2047,
+                40 => 2047,
 #if NETCOREAPP3_0_OR_NEWER
-            return InlineCalculateOffsetMask(Unsafe.SizeOf<T>());
+                _ => InlineCalculateOffsetMask(Unsafe.SizeOf<T>()),
 #else
-            if (Unsafe.SizeOf<T>() == Unsafe.SizeOf<object>())
```
Comment (AlekseyTs) [inline src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs:59]: >if (Unsafe.SizeOf<T>() == Unsafe.SizeOf<object>()) [](http://example.com/codeflow?start=12&length=50)

Here as well.

File: src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs [hunk 6]
```diff
@@ -94,7 +112,7 @@ private static int CalculateSegmentSize(int elementSize)
             static int ArraySize(int elementSize, int segmentSize)
             {
                 // Array object header, plus space for the elements
-                return (2 * IntPtr.Size) + (elementSize * segmentSize);
+                return (2 * IntPtr.Size + 8) + (elementSize * segmentSize);
```
Comment (AlekseyTs) [inline src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs:115]: > + 8 [](http://example.com/codeflow?start=39&length=4)

It is not obvious why this change is necessary.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 1)

File: src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs [hunk 7]
```diff
@@ -19,52 +19,70 @@ internal static class SegmentedArrayHelper
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         internal static int GetSegmentSize<T>()
         {
+            return Unsafe.SizeOf<T>() switch
+            {
+                // Hard code common values since not all versions of .NET support constant-folding the calculation.
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs:24]: This method produces deterministic map from `T` &rarr; constant integer value, but only at runtime. The JIT in .NET 8 is the first version capable of fully calculating this constant value and inlining it at call sites.

File: src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs (hunk 4 above)
Comment (sharwell) [reply src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs:42]: ➡️ Addressing as part of https://github.com/dotnet/roslyn/pull/67921#discussion_r1175772691

File: src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs (hunk 5 above)
Comment (sharwell) [reply src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs:59]: ➡️ Addressing as part of https://github.com/dotnet/roslyn/pull/67921#discussion_r1175772691

File: src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs (hunk 6 above)
Comment (sharwell) [reply src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs:115]: ➡️ Updated the PR commit to mention that this is a follow-up to #67558, where comment https://github.com/dotnet/roslyn/pull/67558#issuecomment-1489441636 called out the incorrect formula

File: src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs (hunk 3 above)
Comment (sharwell) [reply src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs:25]: ➡️ Renamed to `FallbackSegmentHelper<T>`

File: src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (sharwell) [reply src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs]: ➡️ Updated comments to clarify that the constant calculation cannot be performed in the JIT

File: src/Compilers/Core/CodeAnalysisTest/Collections/SegmentedArrayHelperTests.cs (hunk 2 above)
Comment (sharwell) [reply src/Compilers/Core/CodeAnalysisTest/Collections/SegmentedArrayHelperTests.cs:47]: ➡️ Left this code

File: src/Compilers/Core/CodeAnalysisTest/Collections/SegmentedArrayHelperTests.cs [hunk 8]
```diff
@@ -12,6 +17,89 @@ namespace Microsoft.CodeAnalysis.UnitTests.Collections
...
+                yield return new object[] { typeof(Size40) };
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(ExplicitSizeTypes))]
+        public void ExplicitSizesAreCorrect(Type type)
+        {
+            var unsafeSizeOfMethod = typeof(Unsafe).GetMethod(nameof(Unsafe.SizeOf)).MakeGenericMethod(type);
+            var size = (int)unsafeSizeOfMethod.Invoke(null, null);
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/Compilers/Core/CodeAnalysisTest/Collections/SegmentedArrayHelperTests.cs:68]: ➡️ Extracted `InvokeUnsafeSizeOf`

File: src/Compilers/Core/CodeAnalysisTest/Collections/SegmentedArrayHelperTests.cs [hunk 9]
```diff
@@ -12,6 +17,90 @@ namespace Microsoft.CodeAnalysis.UnitTests.Collections
...
```
Comment (AlekseyTs) [inline src/Compilers/Core/CodeAnalysisTest/Collections/SegmentedArrayHelperTests.cs:68]: >(int)unsafeSizeOfMethod.Invoke(null, null); [](http://example.com/codeflow?start=23&length=43)

Use ```InvokeUnsafeSizeOf``` here?

Comment (sharwell) [reply src/Compilers/Core/CodeAnalysisTest/Collections/SegmentedArrayHelperTests.cs:68]: ➡️ Yep, amended the last commit to include this call

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 3)
